From ea73e8752a84685ffd52795ad633938c02ae09b0 Mon Sep 17 00:00:00 2001
From: xvhaocheng <xvhaocheng@gmail.com>
Date: Sat, 17 Oct 2026 14:50:15 +0900
Subject: [PATCH 01/27] Adapt fixed IOMMU ABI with untranslated DMA metadata

Source-organization: 2026-10-17; new commit authored at actual current time.
Feature-ID: CVA6-DTI-COMPAT
Original-source: b98b0ef7
Validation scope: source/patch identity only; no build, simulation or experiment.
No DS delayer, observer, active-cycle counter, E3 or research ports are introduced.
---
 Makefile                                        |  7 ++++---
 corev_apu/fpga/scripts/run.tcl                  |  2 +-
 corev_apu/fpga/src/ariane_peripherals_xilinx.sv |  2 ++
 corev_apu/iDMA/src/idma_backend.sv              |  8 ++++++++
 corev_apu/iDMA/src/idma_legalizer.sv            |  8 ++++++++
 corev_apu/riscv-iommu                           |  2 +-
 corev_apu/tb/ariane_soc_pkg.sv                  | 12 +++++++++++-
 7 files changed, 35 insertions(+), 6 deletions(-)

diff --git a/Makefile b/Makefile
index dd6e2039cf8ed88174298dd81ef63c4d13b298fd..465baa3e37a90be5dd6bd8c0ed4b57d43010ca6c 100644
--- a/Makefile
+++ b/Makefile
@@ -118,9 +118,10 @@ ariane_pkg := \
               corev_apu/tb/ariane_axi_soc_pkg.sv                                         \
               corev_apu/iDMA/src/idma_pkg.sv                                             \
 			  corev_apu/iDMA/src/frontends/register_64bit/idma_reg64_frontend_reg_pkg.sv \
-              corev_apu/riscv-iommu/packages/rv_iommu_pkg.sv                             \
-              corev_apu/riscv-iommu/packages/rv_iommu_field_pkg.sv                       \
-              corev_apu/riscv-iommu/packages/rv_iommu_reg_pkg.sv
+	              corev_apu/riscv-iommu/packages/rv_iommu_pkg.sv                             \
+	              corev_apu/riscv-iommu/packages/rv_iommu_field_pkg.sv                       \
+	              corev_apu/riscv-iommu/packages/rv_iommu_reg_pkg.sv                         \
+	              corev_apu/riscv-iommu/packages/rv_iommu_dti_ats_pkg.sv
               
 ariane_pkg := $(addprefix $(root-dir), $(ariane_pkg))
 
diff --git a/corev_apu/fpga/scripts/run.tcl b/corev_apu/fpga/scripts/run.tcl
index 9f323806a9a781d4bd08b27d7c40ea018a6e241a..b611a82aead000b4793602b799d2af0799e38aee 100644
--- a/corev_apu/fpga/scripts/run.tcl
+++ b/corev_apu/fpga/scripts/run.tcl
@@ -38,7 +38,7 @@ read_ip { \
 }
 # read_ip xilinx/xlnx_protocol_checker/ip/xlnx_protocol_checker.xci
 
-set_property include_dirs { "src/axi_sd_bridge/include" "../../vendor/pulp-platform/common_cells/include" "../../vendor/pulp-platform/axi/include" "../register_interface/include" "../iDMA/src/include"} [current_fileset]
+set_property include_dirs { "src/axi_sd_bridge/include" "../../vendor/pulp-platform/common_cells/include" "../../vendor/pulp-platform/axi/include" "../register_interface/include" "../iDMA/src/include" "../riscv-iommu/include"} [current_fileset]
 
 source scripts/add_sources.tcl
 
diff --git a/corev_apu/fpga/src/ariane_peripherals_xilinx.sv b/corev_apu/fpga/src/ariane_peripherals_xilinx.sv
index 651921485e2c29bd2539a40d973b2c0299dc675f..33ee67104f83e56a0e24404a12cb4d904f980709 100644
--- a/corev_apu/fpga/src/ariane_peripherals_xilinx.sv
+++ b/corev_apu/fpga/src/ariane_peripherals_xilinx.sv
@@ -12,6 +12,8 @@
 
 `include "axi/assign.svh"
 `include "axi/typedef.svh"
+`include "axi-iommu/assign.svh"
+`include "axi-iommu/typedef.svh"
 `include "register_interface/assign.svh"
 `include "register_interface/typedef.svh"
 
diff --git a/corev_apu/iDMA/src/idma_backend.sv b/corev_apu/iDMA/src/idma_backend.sv
index ac162c190a6877860ff025be53ed0f758a3f0ae2..a963bb99cebf6203be4f0ecfb4e6b18b502d5a4c 100644
--- a/corev_apu/iDMA/src/idma_backend.sv
+++ b/corev_apu/iDMA/src/idma_backend.sv
@@ -380,6 +380,10 @@ module idma_backend #(
                 stream_id:      idma_req_i.opt.device_id,
                 ss_id_valid:    idma_req_i.opt.pv,
                 substream_id:   idma_req_i.opt.process_id,
+                mmu_flow:       '0,
+                mmu_atst:       1'b0,
+                mmu_valid:      1'b0,
+                mmu_secsid:     1'b0,
                 addr:           { idma_req_i.src_addr[AddrWidth-1:OffsetWidth], {{OffsetWidth}{1'b0}} },
                 len:            len,
                 size:           axi_pkg::size_t'(OffsetWidth),
@@ -398,6 +402,10 @@ module idma_backend #(
                 stream_id:      idma_req_i.opt.device_id,
                 ss_id_valid:    idma_req_i.opt.pv,
                 substream_id:   idma_req_i.opt.process_id,
+                mmu_flow:       '0,
+                mmu_atst:       1'b0,
+                mmu_valid:      1'b0,
+                mmu_secsid:     1'b0,
                 addr:           { idma_req_i.dst_addr[AddrWidth-1:OffsetWidth], {{OffsetWidth}{1'b0}} },
                 len:            len,
                 size:           axi_pkg::size_t'(OffsetWidth),
diff --git a/corev_apu/iDMA/src/idma_legalizer.sv b/corev_apu/iDMA/src/idma_legalizer.sv
index 10a471a8849fba676ea9b55095c316a94e04b931..3c52d564c4e6d245745afd847c5586f52d06db87 100644
--- a/corev_apu/iDMA/src/idma_legalizer.sv
+++ b/corev_apu/iDMA/src/idma_legalizer.sv
@@ -328,6 +328,10 @@ module idma_legalizer #(
             stream_id:      opt_tf_q.device_id,
             ss_id_valid:    opt_tf_q.pv,
             substream_id:   opt_tf_q.process_id,
+            mmu_flow:       '0,
+            mmu_atst:       1'b0,
+            mmu_valid:      1'b0,
+            mmu_secsid:     1'b0,
             addr:           { r_tf_q.addr[AddrWidth-1:OffsetWidth], {{OffsetWidth}{1'b0}} },
             len:            ((r_num_bytes + r_addr_offset - 'd1) >> OffsetWidth),
             size:           axi_pkg::size_t'(OffsetWidth),
@@ -346,6 +350,10 @@ module idma_legalizer #(
             stream_id:      opt_tf_q.device_id,
             ss_id_valid:    opt_tf_q.pv,
             substream_id:   opt_tf_q.process_id,
+            mmu_flow:       '0,
+            mmu_atst:       1'b0,
+            mmu_valid:      1'b0,
+            mmu_secsid:     1'b0,
             addr:           { w_tf_q.addr[AddrWidth-1:OffsetWidth], {{OffsetWidth}{1'b0}} },
             len:            ((w_num_bytes + w_addr_offset - 'd1) >> OffsetWidth),
             size:           axi_pkg::size_t'(OffsetWidth),
diff --git a/corev_apu/riscv-iommu b/corev_apu/riscv-iommu
index f8f918aa251209a8b7fafde89ba46a01cf0f7c68..7bea3f202903279c18df99a2284d6f56ccdacdc6 160000
--- a/corev_apu/riscv-iommu
+++ b/corev_apu/riscv-iommu
@@ -1 +1 @@
-Subproject commit f8f918aa251209a8b7fafde89ba46a01cf0f7c68
+Subproject commit 7bea3f202903279c18df99a2284d6f56ccdacdc6
diff --git a/corev_apu/tb/ariane_soc_pkg.sv b/corev_apu/tb/ariane_soc_pkg.sv
index 1a4e53c58f767122770370351847b243b1860e30..590cfeac2d3e70c0bb95398b6f274421dc45059e 100644
--- a/corev_apu/tb/ariane_soc_pkg.sv
+++ b/corev_apu/tb/ariane_soc_pkg.sv
@@ -119,6 +119,7 @@ package ariane_soc;
     InclPC              : 1'b1,
     InclAxiBC           : 1'b1,
     InclDbg             : 1'b1,
+    InclATS             : 1'b0,
     MSITrans            : rv_iommu_cfg::MSI_BT_MRIF,
     IGS                 : rv_iommu_cfg::BOTH,
     NumIntVec           : 32'd16,
@@ -128,7 +129,16 @@ package ariane_soc;
     AxiDataWidth        : 32'd64,
     AxiIdWidth          : IdWidth,
     AxiProgIdWidth      : IdWidthSlave,
-    AxiUserWidth        : ariane_pkg::AXI_USER_WIDTH
+    AxiUserWidth        : ariane_pkg::AXI_USER_WIDTH,
+    AxiProIdWidth       : 32'd20,
+    AxiDevIdWidth       : 32'd24,
+    AxisDataWidth       : 32'd160,
+    AxisUserWidth       : 32'd1,
+    AxisKeepWidth       : 32'd20,
+    AxisStrbWidth       : 32'd20,
+    AxisIdWidth         : 32'd1,
+    AxisDestWidth       : 32'd1,
+    Freq                : 32'd100
   };
 
 endpackage
